File: kotlin/java/com/google/ai/edge/litertlm/Session.kt
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.ai.edge.litertlm

import java.util.concurrent.CancellationException
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Manages the lifecycle of a LiteRT-LM session, providing an interface for interacting with the
 * native library.
 *
 * @param handle The pointer to the underlying native session object.
 */
class Session(private val handle: Long) : AutoCloseable {
  private val _isAlive = AtomicBoolean(true)

  /** Whether the session is alive and ready to be used, */
  val isAlive: Boolean
    get() = _isAlive.get()

  /**
   * Closes the session and releases the native session's resources.
   *
   * @throws IllegalStateException if the session has already been closed.
   */
  override fun close() {
    if (_isAlive.compareAndSet(true, false)) {
      LiteRtLmJni.nativeDeleteSession(handle)
    } else {
      throw IllegalStateException("Session is closed already.")
    }
  }

  /**
   * Adds the [InputData] and starts the prefilling process.
   *
   * User can break down their [InputData] into multiple chunks and call this function multiple
   * times.
   *
   * This is a blocking call and the function will return when the prefill process is done.
   *
   * @param inputData An array of [InputData] to be processed by the model.
   * @throws IllegalStateException if the session is not alive.
   */
  fun runPrefill(inputData: List<InputData>) {
    checkIsAlive()
    return LiteRtLmJni.nativeRunPrefill(handle, inputData.toTypedArray())
  }

  /**
   * Runs the decode step for the model to predict the response based on the input data added by
   * [runPrefill].
   *
   * This is a blocking call and the function will return when the decoding process is done.
   *
   * @return The generated content as a [String].
   * @throws IllegalStateException if the session is not alive.
   */
  fun runDecode(): String {
    checkIsAlive()
    return LiteRtLmJni.nativeRunDecode(handle)
  }

  /**
   * Generates content from the provided [InputData] and any previous input data added by
   * [runPrefill].
   *
   * This handles both the prefilling and decoding steps.
   *
   * @param inputData An array of [InputData] to be processed by the model. If the user wants to run
   *   the decode loop only, they can pass an empty array.
   * @return The generated content as a [String].
   * @throws IllegalStateException if the session is not alive.
   */
  fun generateContent(inputData: List<InputData>): String {
    checkIsAlive()
    return LiteRtLmJni.nativeGenerateContent(handle, inputData.toTypedArray())
  }

  /**
   * Generates content from the provided [InputData] and previous input data added by [runPrefill].
   *
   * This handles both the prefilling and decoding steps.
   *
   * @param inputData An array of [InputData] to be processed by the model.
   * @param responseCallback The callback to receive the streaming responses.
   * @throws IllegalStateException if the session is not alive.
   */
  fun generateContentStream(inputData: List<InputData>, responseCallback: ResponseCallback) {
    checkIsAlive()
    val jniCallback = JniInferenceCallbackImpl(responseCallback)
    LiteRtLmJni.nativeGenerateContentStream(handle, inputData.toTypedArray(), jniCallback)
  }

  private inner class JniInferenceCallbackImpl(private val callback: ResponseCallback) :
    LiteRtLmJni.JniInferenceCallback {
    override fun onNext(response: String) {
      callback.onNext(response)
    }

    override fun onDone() {
      callback.onDone()
    }

    override fun onError(statusCode: Int, message: String) {
      if (statusCode == 1) { // StatusCode::kCancelled
        callback.onError(CancellationException(message))
      } else {
        callback.onError(LiteRtLmJniException("Status Code: $statusCode. Message: $message"))
      }
    }
  }

  /**
   * Cancels any ongoing inference process (prefill or decode).
   *
   * If there is no ongoing inference process, it is a no-op.
   *
   * @throws IllegalStateException if the session is not alive.
   */
  fun cancelProcess() {
    checkIsAlive()
    LiteRtLmJni.nativeCancelProcess(handle)
  }

  /** Throws [IllegalStateException] if the session is not alive. */
  private fun checkIsAlive() {
    check(isAlive) { "Session is not alive." }
  }
}

/**
 * A sealed class representing the input data that can be provided to the LiteRT-LM.
 *
 * This corresponds to the native `InputData` in
 * runtime/engine/io_types.h.
 */
sealed class InputData {
  /**
   * Represents text input.
   *
   * @param text The input string.
   */
  data class Text(val text: String) : InputData()

  /**
   * Represents audio input.
   *
   * Supported format: WAV.
   *
   * @param bytes The raw audio data.
   */
  // TODO(b/439003966): add Android-friendly methods to create audio input data.
  data class Audio(val bytes: ByteArray) : InputData()

  /**
   * Represents image input.
   *
   * Supported format: PNG and JPG.
   *
   * @param bytes The raw image data.
   */
  // TODO(b/439003966): add Android-friendly methods to create image input data.
  data class Image(val bytes: ByteArray) : InputData()
}

/** An callback for receiving streaming responses. */
interface ResponseCallback {
  /**
   * Called when a new response is available.
   *
   * @param response The response chunk.
   */
  fun onNext(response: String) {}

  /** Called when the stream is complete. */
  fun onDone() {}

  /**
   * Called when an error occurs.
   *
   * @param throwable The error that occurred. This will be a
   *   [java.util.concurrent.CancellationException] if the stream was cancelled normally, and a
   *   [LiteRtLmJniException] for other errors.
   */
  fun onError(throwable: Throwable) {}
}

File: js/packages/core/src/session.ts
/**
 * Copyright 2026 The ODML Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Mutex} from './mutex.js';
import {SessionConfig, wasmSessionConfigToSessionConfig} from './session_config.js';
import {RecursiveRequired} from './types.js';
import {Responses as WasmResponses, Session as WasmSession} from './wasm_binding_types.js';
import {consumeEmscriptenVectorToArray} from './wasm_utils.js';

const BUSY_MESSAGE = 'Session is busy. An operation is already in progress.';

/**
 * LiteRT-LM Session
 */
export class Session {
  readonly config: RecursiveRequired<SessionConfig>;
  private isBusy = false;

  constructor(private readonly session: WasmSession, private readonly mutexes: {
    executor: Mutex;
  }) {
    const wasmSessionConfig = session.getSessionConfig();
    this.config = wasmSessionConfigToSessionConfig(wasmSessionConfig);
    wasmSessionConfig.delete();
  }

  async runPrefill(inputs: string[]): Promise<void> {
    if (this.isBusy) {
      throw new Error(BUSY_MESSAGE);
    }
    this.isBusy = true;
    try {
      await this.mutexes.executor.acquireAndRun(async () => {
        await this.session.runPrefill(inputs);
      });
    } finally {
      this.isBusy = false;
    }
  }

  async runDecode(): Promise<Responses> {
    if (this.isBusy) {
      throw new Error(BUSY_MESSAGE);
    }
    this.isBusy = true;
    try {
      return await this.mutexes.executor.acquireAndRun(async () => {
        const wasmResponses: WasmResponses = await this.session.runDecode();
        return new Responses(wasmResponses);
      });
    } finally {
      this.isBusy = false;
    }
  }

  /**
   * Sends a signal to cancel any current generation.
   */
  cancel() {
    this.session.cancelProcess();
  }

  /**
   * Clones the session. The cloned session will have all the settings and
   * context of the original session up to the point that clone is called.
   */
  async clone(): Promise<Session> {
    if (this.isBusy) {
      throw new Error(BUSY_MESSAGE);
    }
    this.isBusy = true;
    try {
      return await this.mutexes.executor.acquireAndRun(() => {
        const wasmClonedSession = this.session.clone();
        return new Session(wasmClonedSession, this.mutexes);
      });
    } finally {
      this.isBusy = false;
    }
  }

  async delete(): Promise<void> {
    await this.mutexes.executor.acquireAndRun(() => {
      this.session.delete();
    });
  }
}

class Responses {
  constructor(private readonly responses: WasmResponses) {}

  getTexts(): string[] {
    const texts = this.responses.getTexts();
    return consumeEmscriptenVectorToArray(texts);
  }

  delete() {
    this.responses.delete();
  }
}